### PR TITLE
Prevent progress "leakage" by clearing the sessionStorage progress when we get a DB response

### PR DESCRIPTION
**Pull request by jmkulwik**
Previously, user progress could leak between tabs if a user started their tab sessions in different login states. Specifically:
Tab 1: User is not logged in. User completes levels 1, 2
Tab 2: User logs in. User completes levels 3, 4
User goes back to Tab 1, refreshes, and sees levels 1, 2, 3, and 4 completed
User goes back to Tab 2, refreshes, and sees only levels 3 and 4 completed

This same issue could happen if the user followed the same steps, but with two different accounts.

These two manifestations of this issue are fixed by this change.

Additional details from a similar change: https://github.com/code-dot-org/code-dot-org/pull/29386

Note: A third manifestation of this issue is **not** fixed by this change
Tab 1: User is logged in. User completes levels 1, 2
Tab 2: User logs out.
User goes back to Tab 1, refreshes, and still sees levels 1, 2 completed, despite being logged out.

That will be solved as a [follow-up change](https://codedotorg.atlassian.net/browse/LP-1541) involving one of two options:
1. Update the SessionStorage to use a key/id when a user is logged in and nothing when a user is logged out
2. Update the surrounding code to use SessionStorage when logged out and use the DB when logged in

### Background
Most of these details are recorded in this [slack conversation](https://codedotorg.slack.com/archives/C03CK49G9/p1598478696049400). Copying here for documentation purposes.

SessionStorage is session-specific. That means it is not shared between tabs and is deleted when a tab or browser window is closed.

We use SessionStorage for user progress when the user is logged out. This is because we don't have any information stored in the DB for the user's progress. The SessionStorage is the "source of truth." Once the user logs in, the SessionStorage becomes obsolete. The DB is the new "source of truth." However, we continue to use SessionStorage even once the user logs in for historical reasons (below). This is no longer needed. One option that is now available to us is to only use SessionStorage when the user is logged out.

#### Historical Reasons for SessionStorage

- In Oct 2013, we added support for tracking progress for anonymous users via session[:progress] in Rails controller (using the Rails session cookie): https://github.com/code-dot-org/dashboard/pull/90
- In Sep 2015, various PRs migrated the session-progress tracking out of the Rails session-cookie into an unencrypted progress cookie, with a few related helper functions. This was a transition step making it possible to manipulate the session state from client-side JavaScript in the future: https://github.com/code-dot-org/code-dot-org/pull/4032 https://github.com/code-dot-org/code-dot-org/pull/4060 https://github.com/code-dot-org/code-dot-org/pull/4170 https://github.com/code-dot-org/code-dot-org/pull/4256
- In Oct 2015 (continuing the above work), we moved the progress-tracking logic setting the progress-cookie state from Rails server (in ActivitiesController#track_progress_in_session) to the client (in sendReport): https://github.com/code-dot-org/code-dot-org/pull/4383 https://github.com/code-dot-org/code-dot-org/pull/4452. This made it possible to cache the rendered script-level page for anonymous users and serve it efficiently over CDN, since the set-cookie was no longer being done on the server  response. This also introduced a behavior change to start tracking client-side for logged-in users, since track_progress_in_session was only ever called by anonymous users but sendReport is called by any user.
- In Nov 2015 we started intentionally combining client level progress and UserLevel-based progress when displaying progress for logged-in users: https://github.com/code-dot-org/code-dot-org/pull/5640 ***Based on the comment ('this unblocks async writes') it seems this change was used as a graceful fallback for the HoC scale-mode feature where UserLevel database writes were throttled/queued and persisted asynchronously (hours/days later).***
- In Nov 2015 (a day later) we changed progress-tracking to use sessionStorage instead of a cookie to fix cookie-overflow issues when progress data grows larger than the 4k a cookie can hold: https://github.com/code-dot-org/code-dot-org/pull/5676
- In the future this logic was migrated to React/redux (progress.js linked above) but I think the underlying behavior remained the same. 

The critical line from this description is here: 
> Based on the comment ('this unblocks async writes') it seems this change was used as a graceful fallback for the HoC scale-mode feature where UserLevel database writes were throttled/queued and persisted asynchronously (hours/days later).

This "HoC scale-mode feature" is no longer in use. It was deemed too difficult/complex to maintain.

## Links

- [LP-1471](https://codedotorg.atlassian.net/browse/LP-1471)
- [LP-576](https://codedotorg.atlassian.net/browse/LP-576)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by wjordan** [comment]
#6711 might be interesting/related.
See especially https://github.com/code-dot-org/code-dot-org/pull/6711#discussion_r53102856. If the anonymous multi-tab scenario is being revisited it would make sense to take a look at the original PR and see if its approach is still relevant.
It was reverted (#6952) but the reason unfortunately was not documented.

File: apps/src/code-studio/initApp/loadApp.js [hunk 1]
```diff
@@ -435,6 +442,11 @@ function loadAppAsync(appOptions) {
         }
 
         const store = getStore();
+
+        // Note: We aren't guaranteed that currentUser.signInState will have been set at this point.
+        // It gets set on document.ready. However, it is highly unlikely that the server will return
+        // a response before document.ready is triggered. So far, this hasn't caused problems, so not
+        // worrying about this for now.
```
Comment (jmkulwik) [inline apps/src/code-studio/initApp/loadApp.js:456]: Note: This was just something I noticed while exploring options around using what the client knows about the signed-in user. I didn't end up going down that path for this change, but felt it was worth documenting. If it's confusing here, I can split it into a separate PR, but it seemed benign enough to keep.

**Review by clareconstantine — APPROVED** [review]
This looks good to me - this solution is going to be so much cleaner!

**Review by cforkish — APPROVED** [review]
LGTM! great documentation of the issue.

File: dashboard/app/controllers/api_controller.rb [hunk 2]
```diff
@@ -430,11 +430,8 @@ def user_progress_for_stage
       user_level = current_user.last_attempt(level, script)
       level_source = user_level.try(:level_source).try(:data)
 
-      response[:progress] = current_user.
-        user_levels.
-        by_stage(stage).
-        pluck(:level_id, :best_result).
-        to_h
+      # Temporarily return the full set of progress so we can overwrite what the sessionStorage changed
```
Comment (jmkulwik) [inline dashboard/app/controllers/api_controller.rb:433]: https://codedotorg.atlassian.net/browse/LP-1541

**Pull request merged**